src/server/spatial_map.cpp

```diff
@@ -0,0 +1,172 @@
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "spatial_map.h"
+#include <algorithm>
+
+namespace server
+{
+
+// all inserted entires go into the uncached vector
+void SpatialMap::insert(u16 id, const v3f &pos, bool postIteration)
+{
+	if (m_iterators_stopping_insertion_and_deletion) {
+		m_pending_inserts.insert(SpatialKey(pos, id));
+		return;
+	}
+
+	SpatialKey key(pos);
+	if (postIteration)
+		key = SpatialKey(pos.X, pos.Y, pos.Z, true);
+	m_cached.insert({key, id});
+}
+
+// Invalidates upon position update
+void SpatialMap::updatePosition(u16 id, const v3f &oldPos, const v3f &newPos)
+{
+	// Try to leave early if already in the same bucket:
+	auto range = m_cached.equal_range(SpatialKey(newPos));
+	for (auto it = range.first; it != range.second; ++it) {
+		if (it->second == id) {
+			return; // all good, let's get out of here
+		}
+	}
+
+	remove(id, oldPos); // remove from old cache position
+	insert(id, newPos); // reinsert
+}
+
+void SpatialMap::remove(u16 id, const v3f &pos, bool postIteration)
+{
+	if (m_iterators_stopping_insertion_and_deletion) {
+		m_pending_deletes.insert(SpatialKey(pos, id));
+		return;
+	}
+
+	SpatialKey key(pos);
+	if (postIteration)
+		key = SpatialKey(pos.X, pos.Y, pos.Z, true);
+	if(m_cached.find(key) != m_cached.end()) {
+		auto range = m_cached.equal_range(key);
+		for (auto it = range.first; it != range.second; ++it) {
+			if (it->second == id) {
+				m_cached.erase(it);
+				return; // Erase and leave early
+			}
+		}
+	}
+
+	remove(id); // should never be hit
+}
+
+void SpatialMap::remove(u16 id)
+{
+	if(m_iterators_stopping_insertion_and_deletion) {
+		m_pending_deletes.insert(SpatialKey(v3f(), id));
+		return;
+	}
+
+	for (auto it = m_cached.begin(); it != m_cached.end(); ++it) {
+		if (it->second == id) {
+			m_cached.erase(it);
+			return; // Erase and leave early
+		}
+	}
+}
+
+void SpatialMap::removeAll()
+{
+	if(m_iterators_stopping_insertion_and_deletion) {
+		m_remove_all = true;
+	} else {
+		m_cached.clear();
+	}
+}
+
+void SpatialMap::getRelevantObjectIds(const aabb3f &box, const std::function<void(u16 id)> &callback)
+{
+	if (m_cached.empty()) return;
+
+	// When searching, we must round to maximum extent of relevant mapblock indexes.
+	// Since we're using floats, always assume +-1
+	auto low = [](f32 val) -> s16 {
+		s16 _val = static_cast<s16>(val / BS);
+		return (_val >> 4) - 1;
+	};
+	auto high = [](f32 val) -> s16 {
+		s16 _val = static_cast<s16>(val / BS);
+		return (_val >> 4) + 1;
+	};
+
+	v3s16 min(low(box.MinEdge.X), low(box.MinEdge.Y), low(box.MinEdge.Z)),
+		max(high(box.MaxEdge.X), high(box.MaxEdge.Y), high(box.MaxEdge.Z));
+		
+	// We should only iterate using this spatial map when there are at least 1 objects per mapblocks to check.
+	// Otherwise, might as well just iterate.
+
```

Comment:
if the number of objects is less than 1 per map block than surely the number of objects is small and it doesn't matter much how we check for them?

Reply:
No, actually you can get horrible performance from my spatial map in those cases. Imagine you do an in-radius check within 1000 when there are only 12 active entities. That would be like 1953125 null mapblocks to check, vs iterating over a vector of 12 entities. 

This is my method to ensure we're using the spatial map only when it could possibly benefit us.

Reply:
I wasn't aware this API might be used which such large radii, disregard my comment.